homeassistant/components/esphome/manager.py

```diff
@@ -355,51 +355,97 @@ async def _handle_pipeline_stop(self) -> None:
     async def on_connect(self) -> None:
         """Subscribe to states and list entities on successful API login."""
         entry = self.entry
+        unique_id = entry.unique_id
         entry_data = self.entry_data
         reconnect_logic = self.reconnect_logic
+        assert reconnect_logic is not None, "Reconnect logic must be set"
         hass = self.hass
         cli = self.cli
+        stored_device_name = entry.data.get(CONF_DEVICE_NAME)
+        unique_id_is_mac_address = unique_id and ":" in unique_id
         try:
             device_info = await cli.device_info()
+        except APIConnectionError as err:
+            _LOGGER.warning("Error getting initial data for %s: %s", self.host, err)
+            # Re-connection logic will trigger after this
+            await cli.disconnect()
+            return
 
-            # Migrate config entry to new unique ID if necessary
-            # This was changed in 2023.1
-            if entry.unique_id != format_mac(device_info.mac_address):
-                hass.config_entries.async_update_entry(
-                    entry, unique_id=format_mac(device_info.mac_address)
-                )
+        device_mac = format_mac(device_info.mac_address)
+        mac_address_matches = unique_id == device_mac
+        name_matches = stored_device_name == device_info.name
+        #
+        # Migrate config entry to new unique ID if the current
+        # unique id is not a mac address
+        #
+        # This was changed in 2023.1
+        #
+        if not unique_id_is_mac_address and not mac_address_matches:
+            hass.config_entries.async_update_entry(entry, unique_id=device_mac)
```

Comment:
If the unique id is the old name, we can't do much here as we have to accept the first one we find.  Once its been migrated, the wrong device protection is effective